### ISSUE TITLE
All form fields on a form are deleted when a specific user clicks the save form button

### ISSUE THREAD

**Issue by jamesmacwhite** [verbb/formie#1212]
### Describe the bug

We have encountered a potentially dangerous scenario where by a user with Formie permissions when clicking the save form button seems to trigger all form fields on this form to be removed when it saves.

This isn't happening for all users, so I believe there's a potential user context issue somewhere.

The permissions assigned to the user are the following:

* View forms
* Create forms
* Delete forms
* Manage all forms 
* Manage form appearance 
* Manage form behavior 
* Manage form notifications
* Manage notification advanced 
* Manage notification templates 
* Manage form integrations 
* Manage form settings 

They have access to all forms, rather than individual assignment, although they were assigned the permission for a specific they created

### Steps to reproduce

1. Create new form
2. Drag any fields into the fields tab
3. Use the save form button, after the form is saved all these fields are not saved

This also happens on existing forms, with a potentially dangerous consequence.

1. Select an existing form
2. Use the save button
3. All form fields that were present have been removed from the field layout and deleted (confirmed checking craft_fields) table.

### Form settings

- Multi-page form: No
- Submission Method: N/A
- Client-side Validation: N/A
- Custom Form Templates: No

### Craft CMS version

3.7.61

### Plugin version

1.6.20

### Multi-site?

No

**Comment by jamesmacwhite** [comment]
As a test I set the user temporarily to "Admin" and saving a form did not then delete all the fields, so I don't know if that helps narrow down a potential permissions context issue somewhere.

**Comment by engram-design** [comment]
That seems _incredibly_ strange behaviour, and not something I can reproduce on my end with a few different user permission setups.

Nothing in your logs files? Any other plugins or modules maybe messing around with things? Can you reproduce this if you login on behalf of the user experiencing this?

**Comment by jamesmacwhite** [comment]
It is very strange. I have confirmed it with impersonation as well as the user showing me with their logged in account. It appears to be something specific under their user context though, because my Craft CMS user doesn't do it.

I'll check logs on my local development environment, it was happening across all of our dev, staging and production environments.

The only thing I've caught briefly is a ALTER table command being run with a red error banner, but interestingly Sentry didn't capture it.

I'll look at the logs on my dev with impersonation to see if I can find anything.

**Comment by jamesmacwhite** [comment]
Just so I'm not going mad, this is what is happening, I'm looking for logs now.

https://user-images.githubusercontent.com/8067792/207462959-64e60956-926e-4ca2-bf85-9e6cfdb7b657.mp4

The save action seems to complete, but the fields tab loses all it's fields and they have indeed been deleted, the form table has no data anymore and the fields themselves gone from the craft_fields table.

**Comment by jamesmacwhite** [comment]
It's really strange but, I can't find any logs from Craft or Sentry when the fields are deleted. I have discovered there appears to be something specific to our staging and live environments, running on dev the issue doesn't appear to happen under the user, so this suggests there is something potentially environmental in play. Although, the fact it doesn't happen for my user on our staging and live environments is equally just making it more of a mystery to be honest.

**Comment by jamesmacwhite** [comment]
@engram-design Do you have any idea what potential area this could be related to?

On one test we managed to witness a brief ALTER TABLE red error message banner when the user performed this action, however it wasn't logged in Sentry, so didn't capture the full info.

**Comment by engram-design** [comment]
I've got zero ideas, sorry! I would say to send your database my way, but you mentioned it wasn't replicating on dev.

What's the MySQL versions across your environments?

If you search in all your `storage/logs/*` files for `ALTER TABLE` that might report back the log info?

**Comment by jamesmacwhite** [comment]
I'm going to have to try and parse the raw log files, our staging and live environments are on ephemeral storage with Fortrabbit, so log reading is a bit unconventional.

I do know there's been some recent MySQL changes done to their core infrastructure, so that's a possibility: https://status.fortrabbit.com/notices/nxlma0hfudfkm5sp-immediate-stability-upgrade-for-mysql-eu

MySQL 8 is used across all environments.

The versions specifically in use I'll need to track down in Fortrabbit.

Edit: it looks like based on PHP info on Fortrabbit it's 8.0.30, Nitro dev environment has pulled MySQL 8.0.23.

**Comment by jamesmacwhite** [comment]
Managed to swipe this message performing the save form action on the affected user

```
 Unable to save form: {"general":["SQLSTATE[HY000]: General error: 3507 Failed to update columns dictionary object.\nThe SQL being executed was: ALTER TABLE `craft_fmc_tutoroftheyear` DROP COLUMN `field_ImpactYourNomineeHasHadOnApprenticeDepartment_wxeltxjm`"]}.
```

All the form fields were still gone upon postback.

This might be because of the column name being greater than a certain amount of chars, and not the direct cause.

This does confirm however something is triggering SQL related commands once the save action is made.

Although that unable to save form "general" is a bit weird, unless it's referring to MySQL error general rather than the form handle, as that's not the form handle.

**Comment by jamesmacwhite** [comment]
Further update, analysing the response on the `actions/formie/forms/save` controller. It looks like the fieldLayoutId is being incremented on each save. So it suggests this is a hint of why the fields are disappearing, as it's a new field layout each save by the  looks of it, but again no idea why that's happening.

It's possible that deletion of fields is also triggered by encountering this in the service logic

https://github.com/verbb/formie/blob/craft-3/src/services/Forms.php#L165-L171

But it's more of a side effect of the fieldLayoutId being modified each save.

**Comment by jamesmacwhite** [comment]
https://user-images.githubusercontent.com/8067792/208248865-7077dd7e-7d34-42b6-a397-490c5a219a38.mp4

**Comment by engram-design** [comment]
So that `ALTER TABLE` message is interesting. Usually that means the column name is too long, but 60 characters is under the 64 character limit for table columns.

But yes, it looks like there's an issue saving the field layout, which in turn is causing the fields to be deleted, because they don't belong to a field layout (Formie thinks they've been removed).

Can you create a new form, from scratch but don't do anything to it after the initial save (don't add any fields). Hop into your database and look at the `formie_forms` database table, and ensure there's a value for the `fieldLayoutId` column.

If there isn't, that at least helps us figure out what's going on.

**Comment by jamesmacwhite** [comment]
Performing that test with a new form on our staging environment, the fieldLayoutId was populated on creation without adding any fields, but the same behaviour of when adding fields and clicking save, the value was incremented by one, same behaviour.

The crucial difference here is when doing it on staging, it happens, if I do this test on my development environment, it doesn't happen.

Our website provider Fortrabbit don't seem to think it's anything on there side.

I am going to potentially deploy our website on Servd, to see if I can rule out the environment itself. This is really bizarre and we only migrated to Formie a few months ago, I certainly don't remember this happening.

**Comment by engram-design** [comment]
Yeah, what's strange about all this is that its both environmental and user-specific - but just trying to put that aside.

So the field layout is being created, but not being able to be found next time the form saves with fields.

Maybe you can throw some `Craft::dd()` statements starting [here](https://github.com/verbb/formie/blob/9835774e70db07e247e6b3e12fd8a15b4a9cabb7/src/services/Forms.php#L131) working your way down. Sorry to suggest that - I'd do it for you if I could! Let me know if there's any possibility of gaining access to the CP + filesystem so I can test and not bother you with it.

I'd totally spin up a Fortrabbit site to check, but I don't think that'll be a sure-fire way to replicate the issue either.

**Comment by jamesmacwhite** [comment]
That's the thing, I can't unfortunately, because the app runtime is temporary, I cannot edit the filesystem on our Fortrabbit environment to put in dd/breakpoints. I can on dev of course, but it's not happening on development!

I can give you access to our staging environment, but you won't have access to the filesystem in a conventional way through SSH

**Comment by jamesmacwhite** [comment]
The user specific nature of the error might also be a red herring however, I have been able to have the issue happen on the admin account, although initially it did seem just to be user specific, but that does not appear to be the case, it does look like it's more environmental now.

My only thoughts currently, if there's something to do with our load balancing setup on staging/live e.g. PHP sessions, Redis cache, MySQL redundancy with (master/slave setup) etc.

I have synced all environments to the same MySQL 8 version 8.0.31 (this was recently updated on Fortrabbit) but I don't think that's the issue.

**Comment by engram-design** [comment]
Ah right, fair enough - I'm not an expert with how Fortrabbit works. That does make it a bit tricky to troubleshoot further. Sounds like I might need to spin up a service with them to test.

There was another weird issue with Fortrabbit (unrelated to this one) https://github.com/verbb/formie/issues/822 that I was unable to figure out. I'm wondering if there's just something off about the setup.

**Comment by jamesmacwhite** [comment]
No worries, If you do manage to get in touch to spin up an environment, our staging and live environments run on the professional stack (not universal). You should be able to get a trial from them, you can also mention the organisation name I work for (Nottingham College) and account owner (James White) for reference that you are investigating a potential issue with the Fortrabbit environment and the Formie plugin. I have opened a support case linking to this issue.

I am going to potentially deploy our code to Servd and see if it happens there. Unfortunately, because of needing to verify my work email, the Servd email has been held as potential spam and we don't get the digest emails until Monday to release it, so Servd will have to wait for now!

**Comment by jamesmacwhite** [comment]
The user context area seems to be still in play just to make this more complicated.

My user (james.white) in Craft CMS seems fine, the admin user nottmcoll-admin appears to trigger the constantly changing fieldLayoutId on save and so does another user. I can't see any major differences with the user permissions across the users tested, so I don't know what that's about.

**Comment by jamesmacwhite** [comment]
Just an update I deployed our develop branch of our app on Servd. Admittely a cobbled together environment just so I can access the Control Panel. 

The issue does not occur. So with the fact it doesn't happen on development, and not on Servd with the same codebase, I will have to ask Fortrabbit to look into it some more. It is seemingly the only common factor in all cases.

Although, I think Servd uses MariaDB not MySQL 8, so if it database backend related, then that's not a fair test, but I don't think I can do much about that.

* Doesn't happen on development on any user
* Doesn't happen on staging when deployed on servd (anything targeting environment would trigger)
* Does happen on staging when deployed on Fortrabbit (specific users)
* Does happen on production when deployed on Fortrabbit (specific users)

The user specific nature remains a further mystery, as to why some users can trigger it, others don't.

**Comment by jamesmacwhite** [comment]
Fortrabbit and myself have noticed the following invalid field UID messages, these are appearing when browsing the formie/forms path in Yii Debug toolbar. Not sure if potentially related.

```
yii\base\InvalidArgumentException: Invalid field UID: a9ee2a30-a8bd-46f5-9fa6-769633348ba6 in /app/vendor/craftcms/cms/src/fieldlayoutelements/CustomField.php:109
Stack trace:
#0 /app/vendor/yiisoft/yii2/base/BaseObject.php(159): craft\fieldlayoutelements\CustomField->setFieldUid('a9ee2a30-a8bd-4...')
#1 /app/vendor/yiisoft/yii2/BaseYii.php(558): yii\base\BaseObject->__set('fieldUid', 'a9ee2a30-a8bd-4...')
#2 /app/vendor/yiisoft/yii2/base/BaseObject.php(107): yii\BaseYii::configure(Object(craft\fieldlayoutelements\CustomField), Array)
#3 /app/vendor/craftcms/cms/src/fieldlayoutelements/BaseField.php(59): yii\base\BaseObject->__construct(Array)
#4 /app/vendor/craftcms/cms/src/fieldlayoutelements/CustomField.php(39): craft\fieldlayoutelements\BaseField->__construct(Array)
#5 [internal function]: craft\fieldlayoutelements\CustomField->__construct(NULL, Array)
#6 /app/vendor/yiisoft/yii2/di/Container.php(419): ReflectionClass->newInstanceArgs(Array)
#7 /app/vendor/yiisoft/yii2/di/Container.php(170): yii\di\Container->build('craft\\fieldlayo...', Array, Array)
#8 /app/vendor/yiisoft/yii2/BaseYii.php(365): yii\di\Container->get('craft\\fieldlayo...', Array, Array)
#9 /app/vendor/craftcms/cms/src/Craft.php(61): yii\BaseYii::createObject(Array, Array)
#10 /app/vendor/craftcms/cms/src/services/Fields.php(1362): Craft::createObject(Array)
#11 /app/vendor/craftcms/cms/src/models/FieldLayoutTab.php(141): craft\services\Fields->createLayoutElement(Array)
#12 /app/vendor/yiisoft/yii2/base/BaseObject.php(109): craft\models\FieldLayoutTab->init()
#13 /app/vendor/craftcms/cms/src/services/Fields.php(1247): yii\base\BaseObject->__construct(Array)
#14 /app/vendor/craftcms/cms/src/services/Fields.php(1230): craft\services\Fields->_createLayoutTabFromRow(Array, true)
#15 [internal function]: craft\services\Fields->craft\services\{closure}(Array)
#16 /app/vendor/craftcms/cms/src/services/Fields.php(1229): array_map(Object(Closure), Array)
#17 /app/vendor/craftcms/cms/src/models/FieldLayout.php(252): craft\services\Fields->getLayoutTabsById(5247)
#18 /app/vendor/craftcms/cms/src/models/FieldLayout.php(486): craft\models\FieldLayout->getTabs()
#19 /app/vendor/craftcms/cms/src/services/ElementIndexes.php(332): craft\models\FieldLayout->getCustomFieldElements()
#20 /app/vendor/twig/twig/src/Extension/CoreExtension.php(1566): craft\services\ElementIndexes->getSourceSortOptions('verbb\\formie\\el...', '*')
#21 /app/vendor/craftcms/cms/src/helpers/Template.php(106): twig_get_attribute(Object(craft\web\twig\Environment), Object(Twig\Source), Object(craft\services\ElementIndexes), 'getSourceSortOp...', Array, 'method', false, false)
#22 /app/storage/runtime/compiled_templates/cb/cbb980cdf5a3f1067bf124d917411ae1b2136d174569c15da56ec1522dfd27b2.php(160): craft\helpers\Template::attribute(Object(craft\web\twig\Environment), Object(Twig\Source), Object(craft\services\ElementIndexes), 'getSourceSortOp...', Array, 'method')
#23 /app/vendor/twig/twig/src/Extension/CoreExtension.php(1169): __TwigTemplate_549b4395cd666ef15c17f3b1abb9bfc8e2678d3564118c5810c33f8f91dc1db3->macro_sourceLink('*', Array, true, 'verbb\\formie\\el...')
#24 /app/storage/runtime/compiled_templates/cb/cbb980cdf5a3f1067bf124d917411ae1b2136d174569c15da56ec1522dfd27b2.php(89): twig_call_macro(Object(__TwigTemplate_549b4395cd666ef15c17f3b1abb9bfc8e2678d3564118c5810c33f8f91dc1db3), 'macro_sourceLin...', Array, 53, Array, Object(Twig\Source))
#25 /app/vendor/twig/twig/src/Template.php(405): __TwigTemplate_549b4395cd666ef15c17f3b1abb9bfc8e2678d3564118c5810c33f8f91dc1db3->doDisplay(Array, Array)
#26 /app/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling(Array, Array)
#27 /app/storage/runtime/compiled_templates/2b/2b28259399bdfb093912beedf5b8292ae7e6d65c74e8052ff5d04059641a5cec.php(106): Twig\Template->display(Array)
#28 /app/vendor/twig/twig/src/Template.php(182): __TwigTemplate_2203930ec053efb71421e76cbf07d70c6f7b374dad0bf94fe6029f6ba6feb3e4->block_sidebar(Array, Array)
#29 /app/vendor/twig/twig/src/Template.php(254): Twig\Template->displayBlock('sidebar', Array, Array, true)
#30 /app/storage/runtime/compiled_templates/ed/ed24cdd686642823e5de41ad90ff5849c6fe0898b3865530d739be7bd7d4920a.php(95): Twig\Template->renderBlock('sidebar', Array, Array)
#31 /app/vendor/twig/twig/src/Template.php(405): __TwigTemplate_9b772b97afc4d9dad90501dea4e40ec075b531437759c5889c0d5e748e1fe691->doDisplay(Array, Array)
#32 /app/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling(Array, Array)
#33 /app/storage/runtime/compiled_templates/2b/2b28259399bdfb093912beedf5b8292ae7e6d65c74e8052ff5d04059641a5cec.php(71): Twig\Template->display(Array, Array)
#34 /app/vendor/twig/twig/src/Template.php(405): __TwigTemplate_2203930ec053efb71421e76cbf07d70c6f7b374dad0bf94fe6029f6ba6feb3e4->doDisplay(Array, Array)
#35 /app/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling(Array, Array)
#36 /app/storage/runtime/compiled_templates/73/73539b96d06137e866ece617843a5640917c7a0bc8f2a397448decd377e8b5d5.php(45): Twig\Template->display(Array, Array)
#37 /app/vendor/twig/twig/src/Template.php(405): __TwigTemplate_d7ffab6a8d962418532a0f75c0b88885f8a77d62da4c3bf38ebeece61fd1e9f5->doDisplay(Array, Array)
#38 /app/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling(Array, Array)
#39 /app/storage/runtime/compiled_templates/e9/e934d07d516ad6646878713f075e7fdbdd03ff1e714c9d0a3c1c5413646dfa3b.php(55): Twig\Template->display(Array, Array)
#40 /app/vendor/twig/twig/src/Template.php(405): __TwigTemplate_446571fa7fa22a3c5aa99bc7017df3c976e63b0b3d0a87cd4406d3ad4e54ceef->doDisplay(Array, Array)
#41 /app/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling(Array, Array)
#42 /app/vendor/twig/twig/src/Template.php(390): Twig\Template->display(Array)
#43 /app/vendor/twig/twig/src/TemplateWrapper.php(45): Twig\Template->render(Array, Array)
#44 /app/vendor/twig/twig/src/Environment.php(318): Twig\TemplateWrapper->render(Array)
#45 /app/vendor/craftcms/cms/src/web/View.php(408): Twig\Environment->render('formie/forms/in...', Array)
#46 /app/vendor/craftcms/cms/src/web/View.php(461): craft\web\View->renderTemplate('formie/forms/in...', Array)
#47 /app/vendor/craftcms/cms/src/web/Controller.php(205): craft\web\View->renderPageTemplate('formie/forms/in...', Array, 'cp')
#48 /app/vendor/verbb/formie/src/controllers/FormsController.php(47): craft\web\Controller->renderTemplate('formie/forms/in...', Array)
#49 [internal function]: verbb\formie\controllers\FormsController->actionIndex()
#50 /app/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#51 /app/vendor/yiisoft/yii2/base/Controller.php(178): yii\base\InlineAction->runWithParams(Array)
#52 /app/vendor/yiisoft/yii2/base/Module.php(552): yii\base\Controller->runAction('index', Array)
#53 /app/vendor/craftcms/cms/src/web/Application.php(293): yii\base\Module->runAction('formie/forms/in...', Array)
#54 /app/vendor/yiisoft/yii2/web/Application.php(103): craft\web\Application->runAction('formie/forms/in...', Array)
#55 /app/vendor/craftcms/cms/src/web/Application.php(278): yii\web\Application->handleRequest(Object(craft\web\Request))
#56 /app/vendor/yiisoft/yii2/base/Application.php(384): craft\web\Application->handleRequest(Object(craft\web\Request))
#57 /app/web/index.php(12): yii\base\Application->run()
#58 {main}
```

```
yii\base\InvalidArgumentException: Invalid field UID: 448b1b37-a3a7-4003-887d-83e3d74dcac6 in /app/vendor/craftcms/cms/src/fieldlayoutelements/CustomField.php:109
Stack trace:
#0 /app/vendor/yiisoft/yii2/base/BaseObject.php(159): craft\fieldlayoutelements\CustomField->setFieldUid('448b1b37-a3a7-4...')
#1 /app/vendor/yiisoft/yii2/BaseYii.php(558): yii\base\BaseObject->__set('fieldUid', '448b1b37-a3a7-4...')
#2 /app/vendor/yiisoft/yii2/base/BaseObject.php(107): yii\BaseYii::configure(Object(craft\fieldlayoutelements\CustomField), Array)
#3 /app/vendor/craftcms/cms/src/fieldlayoutelements/BaseField.php(59): yii\base\BaseObject->__construct(Array)
#4 /app/vendor/craftcms/cms/src/fieldlayoutelements/CustomField.php(39): craft\fieldlayoutelements\BaseField->__construct(Array)
#5 [internal function]: craft\fieldlayoutelements\CustomField->__construct(NULL, Array)
#6 /app/vendor/yiisoft/yii2/di/Container.php(419): ReflectionClass->newInstanceArgs(Array)
#7 /app/vendor/yiisoft/yii2/di/Container.php(170): yii\di\Container->build('craft\\fieldlayo...', Array, Array)
#8 /app/vendor/yiisoft/yii2/BaseYii.php(365): yii\di\Container->get('craft\\fieldlayo...', Array, Array)
#9 /app/vendor/craftcms/cms/src/Craft.php(61): yii\BaseYii::createObject(Array, Array)
#10 /app/vendor/craftcms/cms/src/services/Fields.php(1362): Craft::createObject(Array)
#11 /app/vendor/craftcms/cms/src/models/FieldLayoutTab.php(141): craft\services\Fields->createLayoutElement(Array)
#12 /app/vendor/yiisoft/yii2/base/BaseObject.php(109): craft\models\FieldLayoutTab->init()
#13 /app/vendor/craftcms/cms/src/services/Fields.php(1247): yii\base\BaseObject->__construct(Array)
#14 /app/vendor/craftcms/cms/src/services/Fields.php(1230): craft\services\Fields->_createLayoutTabFromRow(Array, true)
#15 [internal function]: craft\services\Fields->craft\services\{closure}(Array)
#16 /app/vendor/craftcms/cms/src/services/Fields.php(1229): array_map(Object(Closure), Array)
#17 /app/vendor/craftcms/cms/src/models/FieldLayout.php(252): craft\services\Fields->getLayoutTabsById(5250)
#18 /app/vendor/craftcms/cms/src/models/FieldLayout.php(486): craft\models\FieldLayout->getTabs()
#19 /app/vendor/craftcms/cms/src/services/ElementIndexes.php(332): craft\models\FieldLayout->getCustomFieldElements()
#20 /app/vendor/twig/twig/src/Extension/CoreExtension.php(1566): craft\services\ElementIndexes->getSourceSortOptions('verbb\\formie\\el...', '*')
#21 /app/vendor/craftcms/cms/src/helpers/Template.php(106): twig_get_attribute(Object(craft\web\twig\Environment), Object(Twig\Source), Object(craft\services\ElementIndexes), 'getSourceSortOp...', Array, 'method', false, false)
#22 /app/storage/runtime/compiled_templates/cb/cbb980cdf5a3f1067bf124d917411ae1b2136d174569c15da56ec1522dfd27b2.php(160): craft\helpers\Template::attribute(Object(craft\web\twig\Environment), Object(Twig\Source), Object(craft\services\ElementIndexes), 'getSourceSortOp...', Array, 'method')
#23 /app/vendor/twig/twig/src/Extension/CoreExtension.php(1169): __TwigTemplate_549b4395cd666ef15c17f3b1abb9bfc8e2678d3564118c5810c33f8f91dc1db3->macro_sourceLink('*', Array, true, 'verbb\\formie\\el...')
#24 /app/storage/runtime/compiled_templates/cb/cbb980cdf5a3f1067bf124d917411ae1b2136d174569c15da56ec1522dfd27b2.php(89): twig_call_macro(Object(__TwigTemplate_549b4395cd666ef15c17f3b1abb9bfc8e2678d3564118c5810c33f8f91dc1db3), 'macro_sourceLin...', Array, 53, Array, Object(Twig\Source))
#25 /app/vendor/twig/twig/src/Template.php(405): __TwigTemplate_549b4395cd666ef15c17f3b1abb9bfc8e2678d3564118c5810c33f8f91dc1db3->doDisplay(Array, Array)
#26 /app/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling(Array, Array)
#27 /app/storage/runtime/compiled_templates/2b/2b28259399bdfb093912beedf5b8292ae7e6d65c74e8052ff5d04059641a5cec.php(106): Twig\Template->display(Array)
#28 /app/vendor/twig/twig/src/Template.php(182): __TwigTemplate_2203930ec053efb71421e76cbf07d70c6f7b374dad0bf94fe6029f6ba6feb3e4->block_sidebar(Array, Array)
#29 /app/vendor/twig/twig/src/Template.php(254): Twig\Template->displayBlock('sidebar', Array, Array, true)
#30 /app/storage/runtime/compiled_templates/ed/ed24cdd686642823e5de41ad90ff5849c6fe0898b3865530d739be7bd7d4920a.php(95): Twig\Template->renderBlock('sidebar', Array, Array)
#31 /app/vendor/twig/twig/src/Template.php(405): __TwigTemplate_9b772b97afc4d9dad90501dea4e40ec075b531437759c5889c0d5e748e1fe691->doDisplay(Array, Array)
#32 /app/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling(Array, Array)
#33 /app/storage/runtime/compiled_templates/2b/2b28259399bdfb093912beedf5b8292ae7e6d65c74e8052ff5d04059641a5cec.php(71): Twig\Template->display(Array, Array)
#34 /app/vendor/twig/twig/src/Template.php(405): __TwigTemplate_2203930ec053efb71421e76cbf07d70c6f7b374dad0bf94fe6029f6ba6feb3e4->doDisplay(Array, Array)
#35 /app/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling(Array, Array)
#36 /app/storage/runtime/compiled_templates/73/73539b96d06137e866ece617843a5640917c7a0bc8f2a397448decd377e8b5d5.php(45): Twig\Template->display(Array, Array)
#37 /app/vendor/twig/twig/src/Template.php(405): __TwigTemplate_d7ffab6a8d962418532a0f75c0b88885f8a77d62da4c3bf38ebeece61fd1e9f5->doDisplay(Array, Array)
#38 /app/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling(Array, Array)
#39 /app/storage/runtime/compiled_templates/e9/e934d07d516ad6646878713f075e7fdbdd03ff1e714c9d0a3c1c5413646dfa3b.php(55): Twig\Template->display(Array, Array)
#40 /app/vendor/twig/twig/src/Template.php(405): __TwigTemplate_446571fa7fa22a3c5aa99bc7017df3c976e63b0b3d0a87cd4406d3ad4e54ceef->doDisplay(Array, Array)
#41 /app/vendor/twig/twig/src/Template.php(378): Twig\Template->displayWithErrorHandling(Array, Array)
#42 /app/vendor/twig/twig/src/Template.php(390): Twig\Template->display(Array)
#43 /app/vendor/twig/twig/src/TemplateWrapper.php(45): Twig\Template->render(Array, Array)
#44 /app/vendor/twig/twig/src/Environment.php(318): Twig\TemplateWrapper->render(Array)
#45 /app/vendor/craftcms/cms/src/web/View.php(408): Twig\Environment->render('formie/forms/in...', Array)
#46 /app/vendor/craftcms/cms/src/web/View.php(461): craft\web\View->renderTemplate('formie/forms/in...', Array)
#47 /app/vendor/craftcms/cms/src/web/Controller.php(205): craft\web\View->renderPageTemplate('formie/forms/in...', Array, 'cp')
#48 /app/vendor/verbb/formie/src/controllers/FormsController.php(47): craft\web\Controller->renderTemplate('formie/forms/in...', Array)
#49 [internal function]: verbb\formie\controllers\FormsController->actionIndex()
#50 /app/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#51 /app/vendor/yiisoft/yii2/base/Controller.php(178): yii\base\InlineAction->runWithParams(Array)
#52 /app/vendor/yiisoft/yii2/base/Module.php(552): yii\base\Controller->runAction('index', Array)
#53 /app/vendor/craftcms/cms/src/web/Application.php(293): yii\base\Module->runAction('formie/forms/in...', Array)
#54 /app/vendor/yiisoft/yii2/web/Application.php(103): craft\web\Application->runAction('formie/forms/in...', Array)
#55 /app/vendor/craftcms/cms/src/web/Application.php(278): yii\web\Application->handleRequest(Object(craft\web\Request))
#56 /app/vendor/yiisoft/yii2/base/Application.php(384): craft\web\Application->handleRequest(Object(craft\web\Request))
#57 /app/web/index.php(12): yii\base\Application->run()
#58 {main}
```

**Comment by jamesmacwhite** [comment]
This is just for more evidence purposes, but testing on our staging environment, same form two different users, save form action. My user (james.white) works as expected, admin user, changes the fieldLayoutId and all fields lost and deleted.

https://user-images.githubusercontent.com/8067792/208786309-c3d76202-2851-4888-8ac1-1a5b1efe4da5.mp4

**Comment by jamesmacwhite** [comment]
After a fun bit of debugging between @engram-design and myself. It has been determined that the default (I think) PHP settings of max_input_vars and potentially post_max_size are the issue on Fortrabbit environments.

It would appear with certain integrations enabled (but not necessarily enabled on a form) this can make the payload significantly large and in this case exceed the set values that Fortrabbit environments had configured.

For reference, my development environment has:

* `max_input_vars` - 5000
* `post_max_size` - 512 MB

Fortrabbit environments:

* `max_input_vars` - 1000
* `post_max_size` - 128 MB

With larger integrations like Dynamics CRM 365 this has made the payload size much bigger and exceeding those values. The reason why it has not been happening on development is because these settings are significantly higher.

Likewise when testing on Servd, I assume those settings were also higher.

Setting the following on Fortrabbit fixes the issue:

* `max_input_vars` - 5000
* `post_max_size` - 256 MB

So ultimately, it's down to the form payload being too large for the request to be handled under the PHP settings.

Hope this helps someone else if they ever end up with this issue!

**Comment by engram-design** [comment]
Glad we got to the bottom of it! I'll see what we can do about excluding that POST data when an integration is disabled.

**Comment by jamesmacwhite** [comment]
Indeed, thanks for looking into it with me!

I recently moved to DDEV and noticed that the post_max_size set is 100M, while max_input_vars is 5000, it seems only max_input_vars is needed to be modified to fix the issue that was occurring going from 1000 to 5000 in this case.

Fortrabbit only allows the values of 64MB 128M and 256M for post_max_size, I have dropped it back to 128M for security, but also made sure DDEV runs with the same value via custom PHP ini so they are in sync, so I can potentially spot this happening across all environments.
